core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Team.hpp

```diff
@@ -0,0 +1,389 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_OPENACC_PARALLEL_REDUCE_TEAM_HPP
+#define KOKKOS_OPENACC_PARALLEL_REDUCE_TEAM_HPP
+
+#include <OpenACC/Kokkos_OpenACC_Team.hpp>
+#include <OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp>
+
+#ifdef KOKKOS_ENABLE_OPENACC_COLLAPSE_HIERARCHICAL_CONSTRUCTS
```

Comment:
This is always compiled for now right? So why do we have an else branch, and how do we know that it compiles? If the else branch is just a "this is what I really would want to write if I had a better compiler" then I'd either not want it in the main branch code, or maybe have it all as a comment?

Reply:
A big (effective) `#if 0` is in many ways a better way to write out a long block of commented code than actual comments. It preserves automatic formatting and display, and about fancier code browser/editor features, in a way that literal comments do not.

Also, it nests, unlike C-style comments

Reply:
@crtrott as you said, the else branch contains the codes that are supposed to be, which may work for other OpenACC compilers (or later version of the NVHPC compiler).

Reply:
I am not so keen on having over 100L of code that has never been compiled and pretend we will just have to flip the switch.  Do we even have an ETA for compiler support?

Commenting as Christian suggested may make it more obvious how the code path may differ.  The code path that is not being compiled is likely buggy for non-sum reductions anyway
```diff
@@ -6,6 +6,9 @@
    using Policy =
        TeamPolicyInternal<Kokkos::Experimental::OpenACC, Properties...>;

+   using ReducerTypeFwd =
+       std::conditional_t<std::is_same_v<InvalidType, ReducerType>, FunctorType,
+                          ReducerType>;
    using Analysis =
        FunctorAnalysis<FunctorPatternInterface::REDUCE, Policy, FunctorType>;
    using value_type   = typename Analysis::value_type;
@@ -26,10 +29,10 @@
      value_type tmp;
      ValueInit::init(a_functor, &tmp);
      if constexpr (!is_reducer_v<ReducerType>) {
- #pragma acc parallel loop gang vector reduction(+ : tmp) copyin(a_functor)
+ #pragma acc parallel loop gang num_gangs(league_size) num_workers(team_size)  vector_length(vector_length) reduction(+:tmp) copyin(a_functor)
        for (int i = 0; i < league_size; i++) {
          int league_id = i;
-         typename Policy::member_type team(league_id, league_size, 1,
+         typename Policy::member_type team(league_id, league_size, team_size,
                                            vector_length);
          a_functor(team, tmp);
        }
@@ -61,20 +64,20 @@
  namespace Kokkos {

  // Hierarchical Parallelism -> Team thread level implementation
- #pragma acc routine seq
+ #pragma acc routine worker
  template <typename iType, class Lambda, typename ValueType>
  KOKKOS_INLINE_FUNCTION std::enable_if_t<!Kokkos::is_reducer<ValueType>::value>
  parallel_reduce(const Impl::TeamThreadRangeBoundariesStruct<
                      iType, Impl::OpenACCTeamMember>& loop_boundaries,
                  const Lambda& lambda, ValueType& result) {
    ValueType tmp = ValueType();
- #pragma acc loop seq
+ #pragma acc loop worker reduction(+ : tmp)
    for (iType i = loop_boundaries.start; i < loop_boundaries.end; i++)
      lambda(i, tmp);
    result = tmp;
  }

- #pragma acc routine seq
+ #pragma acc routine worker
  template <typename iType, class Lambda, typename ReducerType>
  KOKKOS_INLINE_FUNCTION std::enable_if_t<Kokkos::is_reducer<ReducerType>::value>
  parallel_reduce(const Impl::TeamThreadRangeBoundariesStruct<
@@ -83,7 +86,7 @@
    using ValueType = typename ReducerType::value_type;

    ValueType tmp = ValueType();
- #pragma acc loop seq
+ #pragma acc loop worker reduction(+ : tmp)
    for (iType i = loop_boundaries.start; i < loop_boundaries.end; i++)
      lambda(i, tmp);
    result = tmp;
@@ -100,7 +103,7 @@
  }

  // Hierarchical Parallelism -> Thread vector level implementation
- #pragma acc routine seq
+ #pragma acc routine vector
  template <typename iType, class Lambda, typename ValueType>
  KOKKOS_INLINE_FUNCTION void parallel_reduce(
      const Impl::ThreadVectorRangeBoundariesStruct<
@@ -108,14 +111,14 @@
      const Lambda& lambda, ValueType& result) {
    ValueType tmp = ValueType();

- #pragma acc loop seq
+ #pragma acc loop vector reduction(+ : tmp)
    for (iType i = loop_boundaries.start; i < loop_boundaries.end; i++) {
      lambda(i, tmp);
    }
    result = tmp;
  }

- #pragma acc routine seq
+ #pragma acc routine vector
  template <typename iType, class Lambda, typename ReducerType>
  KOKKOS_INLINE_FUNCTION std::enable_if_t<Kokkos::is_reducer<ReducerType>::value>
  parallel_reduce(const Impl::ThreadVectorRangeBoundariesStruct<
@@ -126,7 +129,7 @@
    ValueType tmp;
    ReducerType::init(tmp);

- #pragma acc loop seq
+ #pragma acc loop vector reduction(+ : tmp)
    for (iType i = loop_boundaries.start; i < loop_boundaries.end; i++) {
      lambda(i, tmp);
    }
@@ -144,6 +147,7 @@
  }

  // Hierarchical Parallelism -> Team vector level implementation
+ #pragma acc routine vector
  template <typename iType, class Lambda, typename ValueType>
  KOKKOS_INLINE_FUNCTION void parallel_reduce(
      const Impl::TeamVectorRangeBoundariesStruct<iType, Impl::OpenACCTeamMember>&
@@ -151,7 +155,7 @@
      const Lambda& lambda, ValueType& result) {
    ValueType tmp = ValueType();

- #pragma acc loop seq
+ #pragma acc loop vector reduction(+ : tmp)
    for (iType i = loop_boundaries.start; i < loop_boundaries.end; i++) {
      lambda(i, tmp);
    }

Reply:
I re-factored the whole code so that it can support different types of reductions and reduce the amount of different code paths.